cudamapper/src/chainer_utils.cu

```diff
@@ -0,0 +1,199 @@
+        overlap.target_end_position_in_read_   = start.target_position_in_read_;
+    }
+    else
+    {
+        overlap.relative_strand                = RelativeStrand::Forward;
+        overlap.target_start_position_in_read_ = start.target_position_in_read_;
+        overlap.target_end_position_in_read_   = end.target_position_in_read_;
+    }
+    return overlap;
+}
+
+__global__ void backtrace_anchors_to_overlaps(const Anchor* const anchors,
+                                              Overlap* overlaps,
+                                              double* const scores,
+                                              bool* const max_select_mask,
+                                              int32_t* const predecessors,
+                                              const int64_t n_anchors,
+                                              const int32_t min_score)
+{
+    const int64_t thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+    const int32_t stride    = blockDim.x * gridDim.x;
+
+    for (int i = thread_id; i < n_anchors; i += stride)
+    {
+        if (scores[i] >= min_score)
+        {
+            int32_t index                = i;
+            int32_t first_index          = index;
+            int32_t num_anchors_in_chain = 0;
+            Anchor final_anchor          = anchors[i];
+
+            while (index != -1)
+            {
+                first_index  = index;
+                int32_t pred = predecessors[index];
+                if (pred != -1)
+                {
+                    max_select_mask[pred] = false;
+                }
+                num_anchors_in_chain++;
+                index = predecessors[index];
+            }
+            Anchor first_anchor = anchors[first_index];
+            overlaps[i]         = create_overlap(first_anchor, final_anchor, num_anchors_in_chain);
+        }
+        else
+        {
+            max_select_mask[i] = false;
+        }
+    }
+}
+
+void allocate_anchor_chains(const device_buffer<Overlap>& overlaps,
+                            device_buffer<int32_t>& unrolled_anchor_chains,
+                            device_buffer<int32_t>& anchor_chain_starts,
+                            int64_t& num_total_anchors,
+                            DefaultDeviceAllocator allocator,
+                            cudaStream_t cuda_stream)
+{
+    auto thrust_exec_policy = thrust::cuda::par(allocator).on(cuda_stream);
+    num_total_anchors       = thrust::reduce(thrust_exec_policy,
+                                       thrust::make_transform_iterator(overlaps.begin(), ConvertOverlapToNumResidues()),
+                                       thrust::make_transform_iterator(overlaps.end(), ConvertOverlapToNumResidues()),
+                                       0);
+
+    unrolled_anchor_chains.clear_and_resize(num_total_anchors);
+    anchor_chain_starts.clear_and_resize(overlaps.size());
+
+    thrust::exclusive_scan(thrust_exec_policy,
```

Comment:
Can this be simplified using `thrust::transform_exclusive_scan`?
https://thrust.github.io/doc/group__transformed__prefixsums_gaf67d66dbbff500349ff56671b9ee9c4b.html#gaf67d66dbbff500349ff56671b9ee9c4b

Reply:
Yes, as can the reduce above that. Fixed in 405ec7e4e0b090a3323da5970c038a219078c5bc.